src/NetTopologySuite.IO.ShapeFile/ShapefileDataWriter.cs

```diff
@@ -203,18 +203,28 @@ public void Write(IEnumerable<IFeature> featureCollection)
                     string[] fieldNames = Array.ConvertAll(Header.Fields, field => field.Name);
                     object[] values = new object[fieldNames.Length];
 
+                    int numFeatures = 0;
                     // first, write the one(s) that we scanned already.
                     foreach (var feature in headFeatures)
                     {
                         Write(feature);
+                        numFeatures++;
                     }
 
                     // now continue through the features we haven't scanned yet.
                     while (featuresEnumerator.MoveNext())
                     {
                         Write(featuresEnumerator.Current);
+                        numFeatures++;
                     }
 
+                    // set the number of records
+                    Header.NumRecords = numFeatures;
+                    // Update the header
+                    _dbaseWriter.Write(Header);
+                    // write the end of dbase file marker
+                    _dbaseWriter.WriteEndOfDbf();
```

Comment:
Writing `Header` a second time requires seeking to the beginning to overwrite the correct `NumRecords` and then seek back to the end of the file after, so this would fail if `_dbaseWriter` is backed by a non-seekable stream.

If the current code **ALSO** fails under that same condition, then I don't care.

If it currently works, and this breaks it, then my opinions:
1. `WriteEndOfDbf()` is perfectly fine to do, regardless (though it should probably happen before we seek back to the beginning to finish off the header, since that should be a bit nicer to internal buffers)
1. We should still do as much as possible to deal with this without breaking anyone:
    - .NET 6 added `TryGetNonEnumeratedCount`, which we could [copy-paste on our own](https://github.com/dotnet/runtime/blob/v6.0.4/src/libraries/System.Linq/src/System/Linq/Count.cs#L97-L125) so `numFeatures` is ready **before** we start writing anything, even if the stream isn't seekable.
    - If that fails, but the stream is seekable, then we should be able to do what you have here (but maybe add something that lets us do `bool success = _dbaseWriter.TryWrite(Header, skipIfNotSeekable: true);` so that it can gracefully fall back to the current behavior)
    - If both of those fail (or, well, if they **would** fail, since we have to know in advance), **but the user opts in**, then maybe we can buffer **all** the attribute values to a list before writing **anything** to `_dbaseWriter` so that everything is written forward-only.  If we go that route, it needs to be opt-in, because it could be a dramatic increase in memory usage that someone might be working around some other way.

Again, this big list of things at the end is only there for if the current code **doesn't also fail** if `_dbaseWriter` is backed by a non-seekable stream...

Reply:
Thanks for the feedback. 

Like you said the first one is a non-brainer, but unfortunately the second part with the count is also necessary.
The first solution you provide  `TryGetNonEnumeratedCount` won't help in my case, since it isn't a materialized list.

Concerning the Non Seekable streams, this should already provide issues, when writing a Shp or Shx files.
https://github.com/NetTopologySuite/NetTopologySuite.IO.ShapeFile/blob/develop/src/NetTopologySuite.IO.ShapeFile/ShapefileWriter.cs#L81-L98

With your given remark, I started an investigation how I Could write directly to azure blob storage. This seems to be possible because they added an OpenWrite to the blobclient SDK since 12.11.0, but this is a non seekable stream. This means I won't be able to use this to write shapes

Reply:
>With your given remark, I started an investigation how I Could write directly to azure blob storage. This seems to be possible because they added an OpenWrite to the blobclient SDK since 12.11.0, but this is a non seekable stream. This means I won't be able to use this to write shapes

Have you considered using [flatgeobuf](/flatgeobuf/flatgeobuf)?

Reply:
No I didn't, but I don't think this will solve the issue.
The concrete issue is that the dbase header needs the number of features. Since I'm working with an IEnumerable stream that isn't materialized (completely) in to memory. So I can't call Count on it. 
Since the Stream provided by the OpenWrite in the blobclient doesn't allow seeking (which sounds logical) I can't (re)write the header of the file. 
Most ideal scenario would be to have a performant and cheap way to count all the features.

But this is only the dbase scenario. Shape and Shape Index are even worse. There you need to save the file size in words in side the header.

Reply:
flatgeobuf is optimized for lossless streaming of feature data in both reading and writing.